ydb/core/kqp/opt/physical/effects/kqp_opt_phy_upsert_index.cpp

```diff
@@ -328,30 +330,167 @@ TExprBase MakeUpsertIndexRows(TKqpPhyUpsertIndexMode mode, const TDqPhyPrecomput
     const auto& mainPk = table.Metadata->KeyColumnNames;
     THashSet<TString> usedIndexes;
+    bool hasUniqIndex = false;
     for (const auto& [_, indexDesc] : indexes) {
+        hasUniqIndex |= (indexDesc->Type == TIndexDescription::EType::GlobalSyncUnique);
         for (const auto& indexKeyCol : indexDesc->KeyColumns) {
-            if (inputColumns.contains(indexKeyCol) &&
-                (std::find(mainPk.begin(), mainPk.end(), indexKeyCol) == mainPk.end()))
-            {
-                usedIndexes.insert(indexDesc->Name);
-                break;
+            if (inputColumns.contains(indexKeyCol)) {
+                if (!usedIndexes.contains(indexDesc->Name) &&
+                    std::find(mainPk.begin(), mainPk.end(), indexKeyCol) == mainPk.end())
+                {
+                    usedIndexes.insert(indexDesc->Name);
+                }
+            } else {
+                YQL_ENSURE(std::find(mainPk.begin(), mainPk.end(), indexKeyCol) == mainPk.end());
+                missedKeyInput.emplace(indexKeyCol);
             }
         }
     }
 
-    auto helper = CreateUpsertUniqBuildHelper(table, checkOnlyGivenColumns ? &inputColumns : nullptr, usedIndexes, pos, ctx);
+    if (missedKeyInput && hasUniqIndex) {
+        TVector<TExprBase> columns;
+
+        TCoArgument inLambdaArg(ctx.NewArgument(pos, "in_lambda_arg"));
+        auto missedFromMain = TCoArgument(ctx.NewArgument(pos, "missed_from_main"));
+
+        TVector<TExprBase> resCol;
+
+        for (const auto& x : inputColumns) {
+            if (!missedKeyInput.contains(x)) {
+                resCol.emplace_back(
+                    Build<TCoNameValueTuple>(ctx, pos)
+                        .Name().Build(x)
+                        .Value<TCoMember>()
+                            .Struct(inLambdaArg)
+                            .Name().Build(x)
+                            .Build()
+                        .Done());
+            }
+        }
+
+        TVector<TExprBase> resNullCol = resCol;
+
+        for (const auto& x : missedKeyInput) {
+            auto atom = Build<TCoAtom>(ctx, pos)
+                .Value(x)
+                .Done();
+            columns.emplace_back(atom);
+
+            auto columnType = table.GetColumnType(TString(x));
+            YQL_ENSURE(columnType);
+
+            resCol.emplace_back(
+                    Build<TCoNameValueTuple>(ctx, pos)
+                        .Name().Build(x)
+                        .Value<TCoMember>()
+                            .Struct(missedFromMain)
+                            .Name().Build(x)
+                            .Build()
+                        .Done());
+
+            resNullCol.emplace_back(
+                Build<TCoNameValueTuple>(ctx, pos)
+                    .Name().Build(x)
+                    .Value<TCoNothing>()
+                        .OptionalType(NCommon::BuildTypeExpr(pos, *columnType, ctx))
```

Comment:
I guess this does not work for Pg types, we'll need to somehow construct a pg null here (it corresponds to an empty `TUnboxedValuePod` with correct type)

Reply:
Oh... And make sure we just return correct error in case of not null... Next pr?